lib/mat8bit.gi

```diff
@@ -375,83 +373,38 @@ end);
 
 #  If <ffe> lies in the field of <mat> then we return a matrix in
 #  `Is8BitMatrixRep`, otherwise we delegate to a generic method.
-
-InstallMethod(\*, "for an 8-bit matrix and an internal FFE", IsCollCollsElms,
-[Is8BitMatrixRep, IsFFE and IsInternalRep],
-function(m, s)
-  local q, i, l, r, pv;
-  q := Q_VEC8BIT(m![2]);
-  if not s in GF(q) then
-    TryNextMethod();
-  fi;
-  l := m![1];
-  r := [l];
-  for i in [2 .. l + 1] do
-    pv := m![i] * s;
-    SetFilterObj(pv, IsLockedRepresentationVector);
-    r[i] := pv;
-  od;
-  Objectify(TYPE_MAT8BIT(q, IsMutable(m)), r);
-  return r;
-end);
-
 InstallMethod(\*, "for an 8-bit matrix and an FFE", IsCollCollsElms,
-[Is8BitMatrixRep, IsFFE],
-function(m, s)
-  if IsInternalRep(s) then
-    TryNextMethod();
-  fi;
-  s := AsInternalFFE(s);
-  if s = fail then
-    TryNextMethod();
-  fi;
-  return m * s;
-end);
+[Is8BitMatrixRep, IsFFE], {m, s} -> s * m);
 
 InstallMethod(AdditiveInverseMutable, "for an 8-bit matrix",
 [Is8BitMatrixRep and IsAdditiveElementWithZero],
 function(mat)
   local neg, i, negv;
-  neg := [mat![1]];
-  for i in [2 .. mat![1] + 1] do
-    negv := AdditiveInverseMutable(mat![i]);
+  neg := [NumberRows(mat)];
+  for i in [1 .. NumberRows(mat)] do
+    negv := AdditiveInverseMutable(mat[i]);
     SetFilterObj(negv, IsLockedRepresentationVector);
-    neg[i] := negv;
+    neg[i + 1] := negv;
   od;
-  Objectify(TYPE_MAT8BIT(Q_VEC8BIT(mat![2]), true), neg);
+  Objectify(TYPE_MAT8BIT(Q_VEC8BIT(mat[1]), true), neg);
   return neg;
 end);
 
+# TODO is the next method required? Doesn't use the representation at all
+
 InstallMethod(AdditiveInverseImmutable, "for an 8-bit matrix",
 [Is8BitMatrixRep and IsAdditiveElementWithZero],
-function(mat)
-  local neg, i, negv;
-  neg := [mat![1]];
-  for i in [2 .. mat![1] + 1] do
-    negv := AdditiveInverseImmutable(mat![i]);
-    SetFilterObj(negv, IsLockedRepresentationVector);
-    neg[i] := negv;
-  od;
-  Objectify(TYPE_MAT8BIT(Q_VEC8BIT(mat![2]), false), neg);
-  return neg;
-end);
+mat -> MakeImmutable(AdditiveInverseMutable(mat)));
+
+# TODO is the next method required? Doesn't use the representation at all
```

Comment:
It does use the representation in so far as that it "knows" from the representation that the row objects can be "locked" by setting the `IsLockedRepresentationVector` filter. 

Unfortunately your new code is not equivalent to the old and in particular does not meet the requirements expressed in the documentation for `AdditiveInverseSameMutability`. Specifically: if a matrix is mutable but with immutable rows, then the result of `AdditiveInverseSameMutability` will again be a mutable matrix with immutable rows... I am not saying this is a great API (indeed MatrixObj does away with this kind of oddity), but as long as that's what our documentation says, we should abide by it.

In fact trying to read up on the rules behind all this also lead to me making PR #5263 😂.

Reply:
Aha, ok, this was one of my questions, it wasn't clear to me if this was allowed, some of the code in this file also seems to indicate that it's possible for some rows to be mutable and others to be immutable, is that possible? Interestingly, there don't seem to be any tests that expose the behaviour you (and the doc) describe.